src/js/webrequest.js

```diff
@@ -325,6 +325,65 @@ function blockMozCspReports(details) {
   }
 }
 
+/**
+ * A backstop to remove undesirable redirects for when our JS approach fails.
+ *
+ * Some Google properties (such as Google Docs in Firefox) rewrite clicked
+ * link actions via some mix of capturing all events, window.open() with
+ * document.write(), and meta refresh.
+ *
+ * @param {Object} details webRequest request details object
+ *
+ * @returns {Object|undefined} Can redirect requests
+ */
+function bypassGoogleRedirects(details) {
+  if (!badger.INITIALIZED) { return; }
+  let request_host = extractHostFromURL(details.url);
+  if (!badger.isPrivacyBadgerEnabled(request_host)) { return; }
+  let urlObj, redirect_url;
+  try {
+    urlObj = new URL(details.url);
+  } catch (e) { /* ignore */ }
+  if (urlObj && urlObj.searchParams) {
+    redirect_url = urlObj.searchParams.get('url') || urlObj.searchParams.get('q');
+  }
+  if (redirect_url) {
+    if (redirect_url.startsWith("https://") || redirect_url.startsWith("http://")) {
+      if (utils.isThirdPartyDomain(extractHostFromURL(redirect_url), request_host)) {
+        return {
+          redirectUrl: redirect_url
+        };
+      }
+    }
+  }
+}
+
+/**
+ * Blocks Google's /gen_204 requests.
+ *
+ * @param {Object} details webRequest request details object
+ *
+ * @returns {Object|undefined} Can cancel requests
+ */
+function blockGoogleGen204s(details) {
+  if (!badger.INITIALIZED) { return; }
+
+  let frameData = badger.tabData.getFrameData(details.tabId);
+  if (!frameData) {
+    return;
+  }
```

Comment:
More return collapsing here for readability.
```suggestion
let frameData = badger.tabData.getFrameData(details.tabId);
if (!badger.INITIALIZED || !frameData) { return; }
```

Reply:
Same as above. Tab data may not yet be ready to be read; attempting to read it before Badger is initialized will result in errors.